### DISCUSSION TITLE
GPG Guide

### DISCUSSION THREAD

**Discussion by aceforeverd** [4paradigm/OpenMLDB#289] [Show and tell]
# Summary
This is a tutorial in order to help following [gpg commit convention](https://github.com/4paradigm/rfcs/blob/main/style-guide/commit-convention.md#commit) in [4paradigm/rfcs](https://github.com/4paradigm/rfcs)

# What is gpg

According to the [official website](https://www.gnupg.org/): 
> GnuPG is a complete and free implementation of the [OpenPGP](http://openpgp.org/about/) standard as defined by [RFC4880](https://tools.ietf.org/html/rfc4880) (also known as PGP). GnuPG allows you to encrypt and sign your data and communications; it features a versatile key management system, along with access modules for all kinds of public key directories. GnuPG, also known as GPG, is a command line tool with features for easy integration with other applications. A wealth of frontend applications and libraries are available. GnuPG also provides support for S/MIME and Secure Shell (ssh).

# Install
+ mac: `brew install gnupg`
+ Linux: install `gnupg` from your package manager

# Basic Config

gnupg's default locate at $HOME/.gnupg, you may generate some default config via 
```bash
 addgnupghome username
```
you may interest in the file `~/.gnupg/gpg.conf`, `~/.gnupg/gpg-agent.conf`

# Generate GPG Key

run:
```bash
gpg --full-generate-key
```
follow the instruction and offer your name, email and expire date, default option should be fine. 

**note the email you provided must be one GitHub knows, checkout your emails in [settings/email](https://github.com/settings/emails)**

you can check your gpg key via:
```bash
gpg --list-secret-keys
```

# Export public key
run: 
```bash
gpg --export --armor user-id
```
user-id can be your email or key fingerprint

# Enable GPG in GitHub
go to `settings > SSH and GPG Keys`, copy the public key context and add to gpg keys section.
You may like GitHub's [GPG Tutorial](https://docs.github.com/en/github/authenticating-to-github/managing-commit-signature-verification)

# Enforce GPG Sign in Git
run
```bash
git config --global commit.gpgsign true
```
If you have multiple keys, you can specify the sign key via
```bash
git config --global user.signingkey <key fingerprint>
```
try git commit in your code and push, GitHub should know you :) 
<img width="265" alt="Screen Shot 2021-03-31 at 3 06 34 PM" src="https://user-images.githubusercontent.com/11715200/113104178-cff6a500-9232-11eb-8041-1e30b2733481.png">

# Send/Receive Keys
One important usage of GPG is the [web of trust](https://en.wikipedia.org/wiki/Web_of_trust). You can export you key to gpg servers, then everybody on web can receive your public key without contact directly

to share:
```bash
gpg --send-keys key-id
```
to receive:
```bash
gpg --recv-keys key-id
```
you can pass the extra `--keyserver <server address>` to send/receive key to/from that key server.

# Share your key :-)
aceforeverd: CE4AA0AE5B7E8112

# Upcomming
- [ ] use gpg sign and verify file
- [ ] gpg agent
- [ ] and more

# Some References
+ https://wiki.archlinux.org/index.php/GnuPG
+ https://gnupg.org

**Comment by jingchen2222** [comment]
Chen22: 5EAE89C255C79CC0

**Reply by aceforeverd** [reply]
Not found :(

**Reply by jingchen2222** [reply]
It's weird.
![image](https://user-images.githubusercontent.com/6291173/113325029-b2761980-92cc-11eb-8316-efa6dfbe115c.png)

**Reply by aceforeverd** [reply]
Should send key to a key server first.
```bash
gpg -send-keys user-id
```
then I can find the key via 
```bash
gpg --recv-keys user-id
```

**Reply by imotai** [reply]
imotai:FD3082168701BC57

**Reply by imotai** [reply]
![image](https://user-images.githubusercontent.com/8623385/115991527-7435f680-a5fb-11eb-83c1-7c272b94fe6e.png)

**Reply by jingchen2222** [reply]
> Should send key to a key server first.
> 
> ```shell
> gpg -send-keys user-id
> ```
> 
> then I can find the key via
> 
> ```shell
> gpg --recv-keys user-id
> ```

Pls check again.

**Reply by vagetablechicken** [reply]
HuangWei:C62D6D74980BC645

**Comment by aceforeverd** [comment]
for those who has problem still, maybe checkout external resource like
https://gist.github.com/troyfontaine/18c9146295168ee9ca2b30c00bd1b41e
